File: Blanquita-Backend/app/Models/BobinaPapel/Reportes.py
from datetime import date, datetime, timedelta
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel


class ReporteInventarioBobinaPapelRequest(BaseModel):
    IdsTipoBobina: Optional[list[int]] = None


class ReporteInventarioResumenTipoResponse(BaseModel):
    IdTipoBobina: int
    NombreTipoBobina: str
    CantidadBobinas: int
    PesoNetoTotalKg: Decimal
    GramajePromedio: Decimal
    RecepcionMasAntigua: Optional[date]
    RecepcionMasReciente: Optional[date]


class ReporteInventarioBobinaResponse(BaseModel):
    IdTipoBobina: int
    NombreTipoBobina: str
    IdBobinaPapel: int
    CodigoBobina: str
    CodigoLote: str
    FechaRecepcion: date
    NombreProveedor: str
    PesoBrutoKg: Optional[Decimal]
    PesoNetoKg: Optional[Decimal]
    Gramaje: Optional[Decimal]


class ReporteInventarioBobinaPapelResponse(BaseModel):
    FechaGeneracion: datetime
    TotalBobinas: int
    PesoNetoGeneralKg: Decimal
    Resumen: list[ReporteInventarioResumenTipoResponse]
    Bobinas: list[ReporteInventarioBobinaResponse]


class VerProduccionesBobinaTuboRequest(BaseModel):
    FechaInicio: Optional[date] = None
    FechaFin: Optional[date] = None
    IdTurno: Optional[int] = None
    IdsTipoBobina: Optional[list[int]] = None
    CodigoBobina: Optional[str] = None
    Operador: Optional[str] = None
    IdEstadoProduccion: Optional[int] = None
    Pagina: int = 1
    TamanoPagina: int = 50


class ProduccionBobinaTuboCatalogoResponse(BaseModel):
    IdProduccionBobinaTubo: int
    NombreEstadoProduccion: str
    NombreTurno: str
    Operador: str
    Ci: str
    NombreRol: str
    TipoBobina: str
    CodigoBobina1: str
    CodigoBobina2: str
    FechaInicioProduccion: datetime
    FechaFinProduccion: Optional[datetime]
    DuracionTotal: Optional[timedelta]
    CantidadLogsActual: int


class VerProduccionesBobinaTuboResponse(BaseModel):
    Total: int
    Pagina: int
    TamanoPagina: int
    Producciones: list[ProduccionBobinaTuboCatalogoResponse]


class ReporteProduccionBobinaTuboDetalleRequest(BaseModel):
    IdProduccion: int
    VerPausas: bool = False


class PausaProduccionBobinaTuboResponse(BaseModel):
    IdPausaProduccionBobinaTubo: int
    IdProduccionBobinaTubo: int
    CodigoBobina1: str
    CodigoBobina2: str
    NombreTurno: str
    FechaHoraPausa: datetime
    MotivoPausaProduccion: Optional[str]
    FechaHoraReanudacion: Optional[datetime]
    DuracionPausa: Optional[timedelta]
    OperadorPausa: str
    RolPausa: str
    EstadoPausa: str


class ReporteProduccionBobinaTuboDetalleResponse(BaseModel):
    IdProduccionBobinaTubo: int
    NombreEstadoProduccion: str
    NombreTurno: str
    Operador: str
    Ci: str
    NombreRol: str
    TipoBobina: str
    CodigoBobina1: str
    PesoNeto1: Optional[Decimal]
    Gramaje1: Optional[Decimal]
    Proveedor1: str
    Recepcion1: date
    CodigoBobina2: str
    PesoNeto2: Optional[Decimal]
    Gramaje2: Optional[Decimal]
    Proveedor2: str
    Recepcion2: date
    FechaInicioProduccion: datetime
    FechaFinProduccion: Optional[datetime]
    DuracionTotal: Optional[timedelta]
    CantidadLogsActual: int
    Pausas: Optional[list[PausaProduccionBobinaTuboResponse]] = None
    TotalTiempoPausado: Optional[timedelta] = None


class VerLotesBobinaPapelRequest(BaseModel):
    FechaInicio: Optional[date] = None
    FechaFin: Optional[date] = None
    IdProveedor: Optional[int] = None
    IdsTipoBobina: Optional[list[int]] = None
    Pagina: int = 1
    TamanoPagina: int = 50


class LoteBobinaPapelCatalogoResponse(BaseModel):
    IdLoteBobina: int
    FechaRecepcion: date
    NombreProveedor: str
    CantidadBobinas: int


class VerLotesBobinaPapelResponse(BaseModel):
    Total: int
    Pagina: int
    TamanoPagina: int
    Lotes: list[LoteBobinaPapelCatalogoResponse]


class ReporteLoteBobinaPapelDetalleRequest(BaseModel):
    IdLoteBobina: int


class BobinaLoteDetalleResponse(BaseModel):
    CodigoBobina: str
    NombreTipoBobina: str
    PesoBrutoKg: Optional[Decimal]
    Gramaje: Optional[Decimal]
    PesoNetoKg: Optional[Decimal]


class ReporteLoteBobinaPapelDetalleResponse(BaseModel):
    IdLoteBobina: int
    FechaRecepcion: date
    NombreProveedor: str
    CantidadBobinas: int
    Ci: str
    PrimerNombre: str
    ApellidoPaterno: str
    NombreRol: str
    Bobinas: list[BobinaLoteDetalleResponse]


class ReporteCancelacionProduccionBobinaTuboRequest(BaseModel):
    IdProduccion: int


class CancelacionProduccionBobinaTuboResponse(BaseModel):
    FechaHoraCancelacion: datetime
    MotivoCancelacion: Optional[str]
    Ci: str
    Operador: str
    NombreRol: str


class ReporteCancelacionProduccionBobinaTuboResponse(
    ReporteProduccionBobinaTuboDetalleResponse
):
    Pausas: list[PausaProduccionBobinaTuboResponse]
    TotalTiempoPausado: timedelta
    Cancelacion: CancelacionProduccionBobinaTuboResponse


class ReporteProduccionPorPeriodoRequest(BaseModel):
    FechaInicio: date
    FechaFin: date
    VerCancelaciones: bool = False


class PausaPorMotivoResponse(BaseModel):
    Motivo: str
    CantidadPausas: int
    TiempoTotal: timedelta


class CancelacionPeriodoResponse(BaseModel):
    IdProduccionBobinaTubo: int
    FechaHoraCancelacion: datetime
    MotivoCancelacion: Optional[str]
    Ci: str
    PrimerNombre: str
    ApellidoPaterno: str
    NombreRol: str


class ReporteProduccionPorPeriodoResponse(BaseModel):
    PeriodoInicio: date
    PeriodoFin: date
    TotalProducciones: int
    TotalLogs: int
    Producciones: list[ProduccionBobinaTuboCatalogoResponse]
    PausasPorMotivo: list[PausaPorMotivoResponse]
    TotalPausas: int
    TotalTiempoPausado: timedelta
    Cancelaciones: Optional[list[CancelacionPeriodoResponse]] = None

File: Blanquita-Frontend/src/models/BobinaPapel/Reportes.js
export const ReporteInventarioBobinaPapelRequest = (idsTipoBobina) => ({
  IdsTipoBobina: idsTipoBobina ?? null
});

export const VerProduccionesBobinaTuboRequest = (filtros = {}) => ({
  FechaInicio: filtros.FechaInicio ?? null,
  FechaFin: filtros.FechaFin ?? null,
  IdTurno: filtros.IdTurno ?? null,
  IdsTipoBobina: filtros.IdsTipoBobina ?? null,
  CodigoBobina: filtros.CodigoBobina ?? null,
  Operador: filtros.Operador ?? null,
  IdEstadoProduccion: filtros.IdEstadoProduccion ?? null,
  Pagina: filtros.Pagina ?? 1,
  TamanoPagina: filtros.TamanoPagina ?? 50
});

export const ProduccionBobinaTuboCatalogoResponse = (data) => ({
  IdProduccionBobinaTubo: data.IdProduccionBobinaTubo,
  NombreEstadoProduccion: data.NombreEstadoProduccion,
  NombreTurno: data.NombreTurno,
  Operador: data.Operador,
  Ci: data.Ci,
  NombreRol: data.NombreRol,
  TipoBobina: data.TipoBobina,
  CodigoBobina1: data.CodigoBobina1,
  CodigoBobina2: data.CodigoBobina2,
  FechaInicioProduccion: data.FechaInicioProduccion,
  FechaFinProduccion: data.FechaFinProduccion ?? null,
  DuracionTotal: data.DuracionTotal ?? null,
  CantidadLogsActual: data.CantidadLogsActual
});

export const VerProduccionesBobinaTuboResponse = (data) => ({
  Total: data.Total,
  Pagina: data.Pagina,
  TamanoPagina: data.TamanoPagina,
  Producciones: (data.Producciones ?? []).map(ProduccionBobinaTuboCatalogoResponse)
});

export const ReporteProduccionBobinaTuboDetalleRequest = (idProduccion, verPausas) => ({
  IdProduccion: idProduccion,
  VerPausas: verPausas ?? false
});

export const ReporteCancelacionProduccionBobinaTuboRequest = (idProduccion) => ({
  IdProduccion: idProduccion
});

export const VerLotesBobinaPapelRequest = (filtros = {}) => ({
  FechaInicio: filtros.FechaInicio ?? null,
  FechaFin: filtros.FechaFin ?? null,
  IdProveedor: filtros.IdProveedor ?? null,
  IdsTipoBobina: filtros.IdsTipoBobina ?? null,
  Pagina: filtros.Pagina ?? 1,
  TamanoPagina: filtros.TamanoPagina ?? 50
});

export const LoteBobinaPapelCatalogoResponse = (data) => ({
  IdLoteBobina: data.IdLoteBobina,
  FechaRecepcion: data.FechaRecepcion,
  NombreProveedor: data.NombreProveedor,
  CantidadBobinas: data.CantidadBobinas
});

export const VerLotesBobinaPapelResponse = (data) => ({
  Total: data.Total,
  Pagina: data.Pagina,
  TamanoPagina: data.TamanoPagina,
  Lotes: (data.Lotes ?? []).map(LoteBobinaPapelCatalogoResponse)
});

export const ReporteLoteBobinaPapelDetalleRequest = (idLoteBobina) => ({
  IdLoteBobina: idLoteBobina
});

export const ReporteLotesPorPeriodoRequest = (fechaInicio, fechaFin) => ({
  FechaInicio: fechaInicio,
  FechaFin: fechaFin
});

export const ReporteProduccionPorPeriodoRequest = (fechaInicio, fechaFin, verCancelaciones) => ({
  FechaInicio: fechaInicio,
  FechaFin: fechaFin,
  VerCancelaciones: verCancelaciones ?? false
});

export const VerBobinasPapelRequest = (filtros = {}) => ({
  CodigoBobina: filtros.CodigoBobina ?? null,
  IdProveedor: filtros.IdProveedor ?? null,
  IdsTipoBobina: filtros.IdsTipoBobina ?? null,
  IdEstadoMateriaPrima: filtros.IdEstadoMateriaPrima ?? null,
  IdBobinaPapel: filtros.IdBobinaPapel ?? null,
  Pagina: filtros.Pagina ?? 1,
  TamanoPagina: filtros.TamanoPagina ?? 50
});

export const BobinaPapelCatalogoResponse = (data) => ({
  IdBobinaPapel: data.IdBobinaPapel,
  CodigoBobina: data.CodigoBobina,
  PesoBrutoKg: data.PesoBrutoKg ?? null,
  Gramaje: data.Gramaje ?? null,
  NombreTipoBobina: data.NombreTipoBobina,
  TipoEstado: data.TipoEstado,
  NombreProveedor: data.NombreProveedor
});

export const VerBobinasPapelResponse = (data) => ({
  Total: data.Total,
  Pagina: data.Pagina,
  TamanoPagina: data.TamanoPagina,
  Bobinas: (data.Bobinas ?? []).map(BobinaPapelCatalogoResponse)
});

export const ReporteHistorialMovimientosBobinaRequest = (idBobinaPapel) => ({
  IdBobinaPapel: idBobinaPapel
});
